src/System.Security.Cryptography.Pkcs/src/Internal/Cryptography/Pal/Windows/DecryptorPalWindows.Decrypt.cs

```diff
@@ -17,8 +17,26 @@ namespace Internal.Cryptography.Pal.Windows
 {
     internal sealed partial class DecryptorPalWindows : DecryptorPal
     {
-        public sealed override ContentInfo TryDecrypt(RecipientInfo recipientInfo, X509Certificate2 cert, X509Certificate2Collection originatorCerts, X509Certificate2Collection extraStore, out Exception exception)
+        public sealed override ContentInfo TryDecrypt(RecipientInfo recipientInfo, X509Certificate2 cert, AsymmetricAlgorithm privateKey, X509Certificate2Collection originatorCerts, X509Certificate2Collection extraStore, out Exception exception)
         {
+            Debug.Assert((cert != null) ^ (privateKey != null));
+
+            if (privateKey != null)
+            {
+                ContentInfo contentInfo = _hCryptMsg.GetContentInfo();
+                ContentInfo ret = Helpers.WithKey(AnyOS.ManagedPkcsPal.ManagedKeyTransPal.DecryptCekCore(cert, privateKey as RSA, recipientInfo.EncryptedKey, recipientInfo.KeyEncryptionAlgorithm.Oid.Value, out Exception e),
+                    (cek) =>
+                    {
+                        if (e != null)
+                            return null;
+
+                        return AnyOS.ManagedPkcsPal.ManagedDecryptorPal.TryDecryptCore(cek, contentInfo.ContentType.Value, contentInfo.Content, _contentEncryptionAlgorithm, out e);
```

Comment:
I'm concerned that this flow is going to leave the document in a weird state.  The Windows CMS handle is still tracking an encrypted document.

If you use this flow in PostDecrypt_Encode (StateTests.cs), I'm guessing that the test will fail.

Reply:
@bartonjs do you mean _lastCall field? https://github.com/dotnet/corefx/blob/master/src/System.Security.Cryptography.Pkcs/src/System/Security/Cryptography/Pkcs/EnvelopedCms.cs#L254 If yes then it's being tracked in the managed code and I have changed that part. The specific test you pointed me to is passing when I'm using explicit private key

Reply:
@bartonjs ok, seems like I was setting the state but forgot to validate it